src/csvcubed/utils/cache.py

```diff
@@ -1,4 +1,197 @@
+        try:
+            response = self.http_adapter.send(
+                request,
+                stream=stream,
+                timeout=timeout,
+                verify=verify,
+                cert=cert,
+                proxies=proxies,
+            )
+
+        except requests.exceptions.ConnectionError as e:
+            try:
+                _logger.debug("The connection error has been found")
+
+                path_to_local_file = generate_path_to_local_file(request.url)
+
+                _logger.debug("The local file path is: %s", path_to_local_file)
+
+                return create_local_copy_response(path_to_local_file, request)
+
+            except KeyError:
+                raise FileNotFoundError(
+                    f"URL {request.url} did not produce a response and a local copy could not be found at the corresponding mapped path."
+                ) from e
+
+        if response.status_code >= 400 and response.status_code < 600:
+            _logger.info(f"The status code is {response.status_code}")
+            try:
+                path_to_local_file = generate_path_to_local_file(request.url)
+                return create_local_copy_response(path_to_local_file, request, response)
+
+            except KeyError:
+                raise FileNotFoundError(
+                    f"URL {request.url} produced a invalid response and a local copy could not be found at the corresponding mapped path."
+                )
+
+
+        return response
+
+    def close(self) -> None:
+        self.http_adapter.close()
+
+
+def generate_path_to_local_file(request_url: str) -> Path:
+    trimmed_url = str(request_url).removeprefix("https:")
+    if request_url[len(request_url) - 1] == "/":
+        path_to_local_file = map_url_to_file_path.get(
+            trimmed_url[: len(trimmed_url) - 1]
+        )
+    else:
+        path_to_local_file = map_url_to_file_path.get(trimmed_url)
+
+    if path_to_local_file is None:
+        raise KeyError
+
+    return path_to_local_file
+
+
+def create_local_copy_response(
+    path_to_local_file: Path,
+    request: requests.PreparedRequest,
+    response: Optional[requests.Response] = None,
+) -> requests.Response:
+    """
+    Generates the response object that contains the local copy file path of the requested file and then returns it.
+    """
+    _logger.warning(
+        f"Unable to load json document from given URL. Attempting to load local storage copy of file {path_to_local_file} instead."
+    )
+
+    successful_response = requests.Response()
+
+    successful_response.status_code = 200
+
+    successful_response.raw = BytesIO(bytes(path_to_local_file.read_text(), "utf-8"))
+    successful_response.url = path_to_local_file.as_uri()
+
+    successful_response.encoding = "utf-8"
+
+    if response is not None:
+        successful_response.history = [response]
+
+    successful_response.reason = "OK"
+    successful_response.request = request
+
+    return successful_response
+
+
```

Comment:
Can we move all of this functionality out into a separate file to keep this `cache` file as clean as possible?